File: python/models/NGRC.py
"""
========
model.py
========

Non-linear Vector Autoregressive (NVAR) model from paper:

    Gauthier, D. J., Bollt, E., Griffith, A., & Barbosa, W. A. S. (2021).
    Next generation reservoir computing.
    Nature Communications, 12(1), 5564.
    https://doi.org/10.1038/s41467-021-25801-2

Author: Nathan Trouvain <nathan.trouvain@inria.fr>
Licence: GNU GENERAL PUBLIC LICENSE v3
Copyright (c) 2022 Nathan Trouvain
"""

import itertools as it

import numpy as np
import scipy.linalg
from mip import Model, minimize, xsum


class NGRC:
    def __init__(self, config):
        self.k = config["model"]["params"]["delay"]
        self.s = config["model"]["params"]["stride"]
        self.p = config["model"]["params"]["poly_order"]
        self.alpha = config["model"]["params"]["ridge_param"]
        self.transients = config["model"]["params"]["transients"]
        self.bias = config["model"]["params"]["bias"]
        self.loss = config["model"]["params"]["loss"]
        self.Wout = None
        self.dim = None

    def fit(self, train_data):
        lin_features, nlin_features, _ = self.__nvar(train_data[:-1])
        dTrain = train_data[1:] - train_data[:-1]
        if self.loss == "mse":
            self.Wout = self.__tikhonov_regression(
                lin_features,
                nlin_features,
                dTrain,
            )
        elif self.loss == "mae":
            self.Wout = self.__lp_solve(
                lin_features,
                nlin_features,
                dTrain,
            )
        else:
            raise ValueError(f"Loss {self.loss} not recognized.")

    def predict(self, lin_features, nlin_features):
        """Use the NVAR features and the learned readout matrix for inference.

        Parameters
        ----------
        Wout : numpy.ndarray
            Readout matrix of shape
            (target_dimension, linear_dimension + non_linear_dimension + bias)
        lin_features : numpy.ndarray
            NVAR linear features of shape (timesteps, linear_dimension)
        nlin_features : numpy.ndarray
            NVAR non-linear features of shape (timesteps, nonlinear_dimension)

        Returns
        -------
            numpy.ndarray
                A predicited signal of shape (timesteps, target_dimensions)
        """

        tot_features = np.c_[lin_features, nlin_features]

        # If bias:
        if self.Wout.shape[1] == tot_features.shape[1] + 1:
            W, bias = self.Wout[:, 1:], self.Wout[:, :1]
        else:
            W, bias = self.Wout, np.zeros((self.Wout.shape[0], 1))

        # Transpose to keep time in first axis.
        return (np.dot(W, tot_features.T) + bias).T

    def forecast(self, context, steps):
        _, _, window = self.__nvar(context[-1 * (self.k) - 1 : -1])
        u = np.atleast_2d(context[-1])
        Y = np.zeros((steps, u.shape[1]))
        for i in range(steps):
            lin_features, nlin_features, window = self.__nvar(u, window=window)
            u = u + self.predict(lin_features, nlin_features)
            Y[i, :] = u
        return Y

    def __nvar(self, X, window=None):
        """Apply Non-linear Vecror Autoregressive model (NVAR) to timeseries.

        Parameters
        ----------
        X : numpy.ndarray
            Multivariate timeseries of shape (timesteps, n_dimensions)
        k : int
            Delay of the NVAR
        s : int
            Strides of the NVAR
        p : int
            Order of the non-linear features (i.e. monomials order)

        Returns
        -------
            numpy.ndarray, numpy.ndarray, numpy.ndarray
                Linear features vector, non-linear features vector,
                last sliding window over the signal.
        """
        if self.k < 1:
            raise ValueError("k should be >= 1.")
        if self.s < 1:
            raise ValueError("s should be >= 1.")
        if self.p < 1:
            raise ValueError("p should be > 0.")
        if X.ndim < 2:
            X = X.reshape(-1, 1)

        # Inputs must be of shape (timesteps, dimension)
        n_steps, n_dim = X.shape

        lin_dim = n_dim * self.k  # Linear features dimension

        # Finding all monomials of order p in lin_features requires finding all
        # unique combinations of p lin_features elements, with replacement.
        lin_idx = np.arange(lin_dim)
        monom_idx = np.array(list(it.combinations_with_replacement(lin_idx, self.p)))

        nlin_dim = monom_idx.shape[0]

        # A sliding window to store all lagged inputs, including discarded ones.
        # By default, the window is initialized with zeros,
        # transient features will have unexpected zeros.
        win_dim = (self.k - 1) * self.s + 1  # lagged window dimension
        if window is None:
            window = np.zeros((win_dim, n_dim))
        else:
            if window.shape != (win_dim, n_dim):
                raise ValueError(
                    f"window must be of shape ({win_dim}, {n_dim}) "
                    f"but is of shape {window.shape}."
                )

        # Linear features and non-linear features vectors.
        lin_features = np.zeros((n_steps, lin_dim))
        nlin_features = np.zeros((n_steps, nlin_dim))

        for i in range(n_steps):
            window = np.roll(window, -1, axis=0)
            window[-1, :] = X[i]

            lin_feat = window[:: self.s, :].flatten()
            nlin_feat = np.prod(lin_feat[monom_idx], axis=1)

            lin_features[i, :] = lin_feat
            nlin_features[i, :] = nlin_feat

        return lin_features, nlin_features, window

    def __lp_solve(self, lin_features, nlin_features, target):
        Y = target[self.transients :]
        tot_features = self.__total_feature(lin_features, nlin_features)
        m = Model()
        Wout_var = m.add_var_tensor(
            (Y.shape[1], tot_features.shape[1]), lb=-5.0, ub=5.0
        )

        z = m.add_var_tensor(Y.shape, lb=0.0)
        for i in range(z.shape[0]):
            for j in range(z.shape[1]):
                m += (
                    Y[i, j]
                    - xsum(
                        Wout_var[j, k] * tot_features[i, k]
                        for k in range(tot_features.shape[1])
                    )
                    <= z[i, j]
                )
                m += (
                    Y[i, j]
                    - xsum(
                        Wout_var[j, k] * tot_features[i, k]
                        for k in range(tot_features.shape[1])
                    )
                    >= -z[i, j]
                )
        m.objective = minimize(
            xsum(z[i, j] for i in range(z.shape[0]) for j in range(z.shape[1]))
        )
        m.optimize()
        Wout = Wout_var.astype(float).to_numpy()
        return Wout

    def __tikhonov_regression(self, lin_features, nlin_features, target):
        """Performs Tikhonov linear regression (with L2 regularization) between
        NVAR features and a target signal to create a readout weight matrix.

        Parameters
        ----------
        lin_features : numpy.ndarray
            NVAR linear features of shape (timesteps, linear_dimension)
        nlin_features : numpy.ndarray
            NVAR non-linear features of shape (timesteps, nonlinear_dimension)
        target : numpy.ndarray
            Target signal, of shape (timesteps, target_dimension)
        alpha : float
            Regularization coefficient
        transients : int
            Number of timesteps to consider as transients (will be discarded before
            linear regression)
        bias : bool
            If True, add a constant term to NVAR features to compute intercept
            during linear regression

        Returns
        -------
            numpy.ndarray
                Readout weights matrix of shape
                (target_dimension, linear_dimension + non_linear_dimension + bias)
        """
        Y = target[self.transients :]

        if Y.ndim < 2:
            Y = Y.reshape(-1, 1)

        tot_features = self.__total_feature(lin_features, nlin_features)

        Wout = np.zeros((Y.shape[1], tot_features.shape[1]))

        # Wout = Y.Otot^T.(Otot.Otot^T + alphaId)^-1
        # (inverted all transpose as we prefer having time on the first axis)
        YXt = np.dot(Y.T, tot_features)
        XXt = np.dot(tot_features.T, tot_features)
        ridge = self.alpha * np.identity(len(XXt), dtype=np.float64)

        Wout[:] = np.dot(YXt, scipy.linalg.pinvh(XXt + ridge))

        return Wout

    def __total_feature(self, lin_features, nlin_features):

        n_steps = len(lin_features) - self.transients

        tot_features = np.c_[lin_features, nlin_features][self.transients :]

        if self.bias:
            c = np.ones((n_steps, 1))
            tot_features = np.c_[c, tot_features]

        return np.c_[lin_features, nlin_features]

File: main/src/models/ngrc.rs
#![allow(dead_code)]

use ndarray::{s, Array1, Array2, Axis};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum NgrcLoss {
    Mse,
    Mae,
}

#[derive(Debug, Clone)]
pub struct NgrcConfig {
    pub delay: usize,
    pub stride: usize,
    pub poly_order: usize,
    pub ridge_param: f32,
    pub transients: usize,
    pub bias: bool,
    pub loss: NgrcLoss,
}

impl Default for NgrcConfig {
    fn default() -> Self {
        Self {
            delay: 2,
            stride: 1,
            poly_order: 2,
            ridge_param: 1e-6,
            transients: 0,
            bias: true,
            loss: NgrcLoss::Mse,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Ngrc {
    pub config: NgrcConfig,
    pub wout: Option<Array2<f32>>,
}

impl Ngrc {
    pub fn new(config: NgrcConfig) -> Self {
        Self { config, wout: None }
    }

    pub fn fit(&mut self, train_data: &Array2<f32>) -> Result<(), String> {
        if train_data.nrows() < 2 {
            return Err("train_data must have at least 2 timesteps".to_string());
        }

        let x = train_data.slice(s![..-1, ..]).to_owned();
        let y_next = train_data.slice(s![1.., ..]).to_owned();
        let dtrain = &y_next - &x;

        let (lin_features, nlin_features, _) = self.nvar(&x, None)?;
        let wout = match self.config.loss {
            NgrcLoss::Mse => self.tikhonov_regression(&lin_features, &nlin_features, &dtrain)?,
            // LP-based MAE in Python requires an LP solver; here we fallback to ridge for portability.
            NgrcLoss::Mae => self.tikhonov_regression(&lin_features, &nlin_features, &dtrain)?,
        };
        self.wout = Some(wout);
        Ok(())
    }

    pub fn forecast(&self, context: &Array2<f32>, steps: usize) -> Result<Array2<f32>, String> {
        if context.nrows() < self.config.delay + 1 {
            return Err(format!(
                "context is too short: need at least {} rows",
                self.config.delay + 1
            ));
        }
        let n_dim = context.ncols();
        let mut output = Array2::<f32>::zeros((steps, n_dim));

        let warmup = context
            .slice(s![
                context.nrows() - self.config.delay - 1..context.nrows() - 1,
                ..
            ])
            .to_owned();
        let (_, _, mut window) = self.nvar(&warmup, None)?;

        let mut u = context
            .slice(s![context.nrows() - 1..context.nrows(), ..])
            .to_owned();

        for i in 0..steps {
            let (lin, nlin, new_window) = self.nvar(&u, Some(window))?;
            window = new_window;
            let delta = self.predict_from_features(&lin, &nlin)?;
            u = &u + &delta;
            output.slice_mut(s![i..i + 1, ..]).assign(&u);
        }

        Ok(output)
    }

    fn predict_from_features(
        &self,
        lin_features: &Array2<f32>,
        nlin_features: &Array2<f32>,
    ) -> Result<Array2<f32>, String> {
        let Some(wout) = &self.wout else {
            return Err("model is not fitted".to_string());
        };
        let mut tot = ndarray::concatenate![Axis(1), lin_features.view(), nlin_features.view()];
        if self.config.bias {
            let ones = Array2::<f32>::ones((tot.nrows(), 1));
            tot = ndarray::concatenate![Axis(1), ones.view(), tot.view()];
        }
        Ok(tot.dot(&wout.t()))
    }

    fn nvar(
        &self,
        x: &Array2<f32>,
        window: Option<Array2<f32>>,
    ) -> Result<(Array2<f32>, Array2<f32>, Array2<f32>), String> {
        let k = self.config.delay;
        let s_stride = self.config.stride;
        let p = self.config.poly_order;

        if k < 1 || s_stride < 1 || p < 1 {
            return Err("delay/stride/poly_order must be >= 1".to_string());
        }

        let n_steps = x.nrows();
        let n_dim = x.ncols();
        let lin_dim = n_dim * k;

        let monom_idx = combinations_with_replacement(lin_dim, p);
        let nlin_dim = monom_idx.len();

        let win_dim = (k - 1) * s_stride + 1;
        let mut win = match window {
            Some(w) => {
                if w.shape() != [win_dim, n_dim] {
                    return Err(format!(
                        "window shape mismatch: expected ({}, {}), got ({}, {})",
                        win_dim,
                        n_dim,
                        w.nrows(),
                        w.ncols()
                    ));
                }
                w
            }
            None => Array2::<f32>::zeros((win_dim, n_dim)),
        };

        let mut lin_features = Array2::<f32>::zeros((n_steps, lin_dim));
        let mut nlin_features = Array2::<f32>::zeros((n_steps, nlin_dim));

        for i in 0..n_steps {
            for row in 0..(win_dim - 1) {
                let next = win.slice(s![row + 1, ..]).to_owned();
                win.slice_mut(s![row, ..]).assign(&next);
            }
            win.slice_mut(s![win_dim - 1, ..])
                .assign(&x.slice(s![i, ..]).to_owned());

            let mut lin_feat = Array1::<f32>::zeros(lin_dim);
            let mut idx = 0usize;
            for row in (0..win_dim).step_by(s_stride) {
                for col in 0..n_dim {
                    lin_feat[idx] = win[[row, col]];
                    idx += 1;
                }
            }

            let mut nlin_feat = Array1::<f32>::zeros(nlin_dim);
            for (j, comb) in monom_idx.iter().enumerate() {
                let mut prod = 1.0f32;
                for &c in comb {
                    prod *= lin_feat[c];
                }
                nlin_feat[j] = prod;
            }

            lin_features.slice_mut(s![i, ..]).assign(&lin_feat);
            nlin_features.slice_mut(s![i, ..]).assign(&nlin_feat);
        }

        Ok((lin_features, nlin_features, win))
    }

    fn total_feature(
        &self,
        lin_features: &Array2<f32>,
        nlin_features: &Array2<f32>,
    ) -> Array2<f32> {
        let mut tot = ndarray::concatenate![Axis(1), lin_features.view(), nlin_features.view()];
        let trans = self.config.transients.min(tot.nrows());
        tot = tot.slice(s![trans.., ..]).to_owned();
        if self.config.bias {
            let ones = Array2::<f32>::ones((tot.nrows(), 1));
            tot = ndarray::concatenate![Axis(1), ones.view(), tot.view()];
        }
        tot
    }

    fn tikhonov_regression(
        &self,
        lin_features: &Array2<f32>,
        nlin_features: &Array2<f32>,
        target: &Array2<f32>,
    ) -> Result<Array2<f32>, String> {
        let trans = self.config.transients.min(target.nrows());
        let y = target.slice(s![trans.., ..]).to_owned();
        let x = self.total_feature(lin_features, nlin_features);

        if x.nrows() == 0 || y.nrows() == 0 {
            return Err("no rows left after applying transients".to_string());
        }

        let yxt = y.t().dot(&x);
        let mut xxt = x.t().dot(&x);
        for i in 0..xxt.nrows().min(xxt.ncols()) {
            xxt[[i, i]] += self.config.ridge_param;
        }
        let inv = invert_square(&xxt)?;
        Ok(yxt.dot(&inv))
    }
}

fn combinations_with_replacement(n: usize, p: usize) -> Vec<Vec<usize>> {
    fn rec(start: usize, n: usize, p: usize, cur: &mut Vec<usize>, out: &mut Vec<Vec<usize>>) {
        if cur.len() == p {
            out.push(cur.clone());
            return;
        }
        for i in start..n {
            cur.push(i);
            rec(i, n, p, cur, out);
            cur.pop();
        }
    }

    let mut out = Vec::new();
    rec(0, n, p, &mut Vec::new(), &mut out);
    out
}

fn invert_square(a: &Array2<f32>) -> Result<Array2<f32>, String> {
    if a.nrows() != a.ncols() {
        return Err("matrix must be square".to_string());
    }
    let n = a.nrows();
    let mut aug = Array2::<f32>::zeros((n, 2 * n));
    for r in 0..n {
        for c in 0..n {
            aug[[r, c]] = a[[r, c]];
        }
        aug[[r, n + r]] = 1.0;
    }

    for i in 0..n {
        let mut pivot = i;
        let mut max_abs = aug[[i, i]].abs();
        for r in (i + 1)..n {
            let v = aug[[r, i]].abs();
            if v > max_abs {
                max_abs = v;
                pivot = r;
            }
        }
        if max_abs < 1e-12 {
            return Err("matrix is singular".to_string());
        }
        if pivot != i {
            for c in 0..(2 * n) {
                let tmp = aug[[i, c]];
                aug[[i, c]] = aug[[pivot, c]];
                aug[[pivot, c]] = tmp;
            }
        }

        let diag = aug[[i, i]];
        for c in 0..(2 * n) {
            aug[[i, c]] /= diag;
        }

        for r in 0..n {
            if r == i {
                continue;
            }
            let factor = aug[[r, i]];
            if factor.abs() < 1e-20 {
                continue;
            }
            for c in 0..(2 * n) {
                aug[[r, c]] -= factor * aug[[i, c]];
            }
        }
    }

    let mut inv = Array2::<f32>::zeros((n, n));
    for r in 0..n {
        for c in 0..n {
            inv[[r, c]] = aug[[r, n + c]];
        }
    }
    Ok(inv)
}

#[cfg(test)]
mod tests {
    use super::*;
    use ndarray::Array2;

    #[test]
    fn ngrc_fit_and_forecast_shape() {
        let mut model = Ngrc::new(NgrcConfig {
            delay: 2,
            stride: 1,
            poly_order: 2,
            ridge_param: 1e-4,
            transients: 5,
            bias: true,
            loss: NgrcLoss::Mse,
        });

        let n = 120;
        let mut data = Array2::<f32>::zeros((n, 1));
        for i in 0..n {
            data[[i, 0]] = (i as f32 * 0.05).sin();
        }

        model.fit(&data).unwrap();
        let context = data.slice(s![n - 30.., ..]).to_owned();
        let pred = model.forecast(&context, 10).unwrap();
        assert_eq!(pred.shape(), &[10, 1]);
    }
}
